### DISCUSSION TITLE
Deploying Prisma with Nest.js to Google App Engine (GAE):  @prisma/client did not initialize yet.

### DISCUSSION THREAD

**Discussion by areindl** [prisma/prisma#10156] [Q&A]
Hi all!

I am currently facing problems deploying Prisma to Google App Engine (GAE). I am using Nest.js and have created a custom module to make sure that Prisma is only initialized once:

```js

import { Global, INestApplication, OnModuleInit,  } from '@nestjs/common'
import { PrismaClient } from '@prisma/client'


@Global()

export class PrismaService extends PrismaClient implements OnModuleInit {
  async onModuleInit() {
    await this.$connect()
  }

  async enableShutdownHooks(app: INestApplication) {
    this.$on('beforeExit', async () => {
      await app.close()
    })
  }
}

```

Here is my package.json

```json
{
  "name": "my-api",
  "version": "1.0.0",
  "description": "",
  "author": "",
  "private": true,
  "license": "UNLICENSED",
  "main": "dist/src/main.js",
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "tsc -p tsconfig.build.json",
    "gcp-build": "npm run build",
    "deploy": "gcloud app deploy -q",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "node ./dist/src/main.js",
    "start:dev": "DEPLOYMENT_TARGET=local nest start --watch",
    "start:debug": "DEPLOYMENT_TARGET=local nest start --debug --watch",
    "serve": "npm run start:dev",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --config ./test/jest-e2e.json",
    "postinstall": "prisma generate"
  },
  "dependencies": {
    "@google-cloud/debug-agent": "^5.2.8",
    "@nestjs/common": "^8.0.0",
    "@nestjs/config": "^1.0.2",
    "@nestjs/core": "^8.0.0",
    "@nestjs/passport": "^8.0.1",
    "@nestjs/platform-express": "^8.0.0",
    "@nestjs/swagger": "^5.1.4",
    "@prisma/client": "^3.4.1",
    "bunyan": "^1.8.15",
    "class-transformer": "^0.4.0",
    "class-validator": "^0.13.1",
    "firebase-admin": "^9.12.0",
    "passport-firebase-jwt": "^1.2.1",
    "pg": "^8.7.1",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^7.2.0",
    "swagger-ui-express": "^4.1.6"
  },
  "devDependencies": {
    "@google-cloud/nodejs-repo-tools": "^3.3.0",
    "@nestjs/cli": "^8.0.0",
    "@nestjs/schematics": "^8.0.0",
    "@nestjs/testing": "^8.0.0",
    "@types/express": "^4.17.13",
    "@types/jest": "^27.0.1",
    "@types/node": "^16.0.0",
    "@types/supertest": "^2.0.11",
    "@typescript-eslint/eslint-plugin": "^4.28.2",
    "@typescript-eslint/parser": "^4.28.2",
    "@vegardit/prisma-generator-nestjs-dto": "^1.4.1",
    "eslint": "^7.30.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^3.4.0",
    "jest": "^27.0.6",
    "prettier": "^2.3.2",
    "prisma": "^3.4.1",
    "supertest": "^6.1.3",
    "ts-jest": "^27.0.3",
    "ts-loader": "^9.2.3",
    "ts-node": "^10.0.0",
    "tsconfig-paths": "^3.10.1",
    "typescript": "^4.3.5"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  },
  "prisma": {
    "seed": "ts-node prisma/seed.ts"
  }
}

```

I looked at https://github.com/prisma/prisma/discussions/10149 and added `"postinstall": "prisma generate"`.

I also tried to to run the generate commanded before the build process. did not help.

Any clue? Did someone solve that?

I am still seeing the error on LogExplorer:

![image](https://user-images.githubusercontent.com/5536491/140738347-1011d156-346c-4442-ad4f-0a239c6e7d3b.png)

Thanks in advance!

**Comment by nikolasburk** [comment]
Hey @areindl, thanks so much for sharing this error here! I've forwarded this to our Engineering team by creating an issue so that they can try to reproduce the problem: https://github.com/prisma/prisma/issues/10163

On a related note: We track how to deploy apps that use Prisma to various deployment platforms [here](https://github.com/prisma/e2e-tests). We don't have a test for GAE yet but are definitely considering to add one once we have this figured out!

**Reply by areindl** [reply]
Update:

I start to believe it is not possible with GAE.

According to the docs:
> 
> 
> 
> "At deployment, before starting your application, you can perform a custom build step by adding a gcp-build script in your package.json file.
> 
> When this script is executed, the dependencies in the dependencies and devDependencies fields of your package.json file are available. After executing your custom build step, App Engine removes and regenerates the node_modules folder by only installing the production dependencies declared in the dependencies field of your package.json file."

Source: https://cloud.google.com/appengine/docs/standard/nodejs/running-custom-build-step

That means: GAE regenerates `node_modules` which is why it is not initialized - although I ran `generate` before.

I will try using a docker-based solution.

**Reply by jonahallibone** [reply]
You could try and output the binary [somewhere else](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#using-a-custom-output-path) during your custom build step?

**Comment by worktechnotoil** [comment]
i am also facing the same issue .

**Comment by naquiroz** [comment]
For anyone still with this issue, I believe it can be solved by generating the client somewhere else:
https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#the-location-of-prisma-client

**Comment by torver213** [comment]
If your still looking for the solution in 2024, i got your back. Checkout this repo with the full implementation of [Deploying a TypeScript Node.js, PostgreSQL & Prisma App on Google Cloud Platform (GCP) App Engine Flexible Environment with SSE Implementation](https://github.com/torver213/gcp-prisma-sse)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
